src/cpp/subscriber/SubscriberHistory.cpp

```diff
@@ -77,300 +90,226 @@ bool SubscriberHistory::received_change(
-                add = true;
-            }
-        }
-        else if (m_historyQos.kind == KEEP_LAST_HISTORY_QOS)
-        {
-            if (m_changes.size() < (size_t)m_historyQos.depth)
-            {
-                add = true;
-            }
-            else
-            {
-                // TODO (Ricardo) Older samples should be selected by sourcetimestamp.
+        return add_received_change(a_change);
+    }
 
-                // Try to substitute a older samples.
-                CacheChange_t* older = nullptr;
+    return false;
+}
 
-                for (auto it = m_changes.begin(); it != m_changes.end(); ++it)
-                {
-                    if ((*it)->writerGUID == a_change->writerGUID &&
-                        (*it)->sequenceNumber < a_change->sequenceNumber)
-                    {
-                        older = *it;
-                        break;
-                    }
-                }
+bool SubscriberHistory::received_change_keep_last_no_key(
+        CacheChange_t* a_change,
+        size_t /* unknown_missing_changes_up_to */ )
+{
+    bool add = false;
+    if (m_changes.size() < static_cast<size_t>(m_historyQos.depth) )
+    {
+        add = true;
+    }
+    else
+    {
+        // Try to substitute the oldest sample.
 
-                if (older != nullptr)
-                {
-                    bool read = older->isRead;
+        // As the history should be ordered following the presentation QoS, we can always remove the first one.
+        add = remove_change_sub(m_changes.at(0));
+    }
 
-                    if (this->remove_change_sub(older))
-                    {
-                        if (!read)
-                        {
-                            this->decreaseUnreadCount();
-                        }
-                        add = true;
-                    }
-                }
-            }
-        }
+    if (add)
+    {
+        return add_received_change(a_change);
+    }
 
-        if (add)
-        {
-            if (m_isHistoryFull)
-            {
-                // Discarting the sample.
-                logWarning(SUBSCRIBER, "Attempting to add Data to Full ReaderHistory: " << this->mp_subImpl->getGuid().entityId);
-                return false;
-            }
+    return false;
+}
 
-            if (this->add_change(a_change))
-            {
-                increaseUnreadCount();
-                if ((int32_t)m_changes.size() == m_resourceLimitsQos.max_samples)
-                    m_isHistoryFull = true;
-                logInfo(SUBSCRIBER, this->mp_subImpl->getGuid().entityId
-                    << ": Change " << a_change->sequenceNumber << " added from: "
-                    << a_change->writerGUID;);
+bool SubscriberHistory::received_change_keep_all_with_key(
+        CacheChange_t* a_change,
+        size_t /* unknown_missing_changes_up_to */ )
+{
+    // TODO(Miguel C): Should we check unknown_missing_changes_up_to as it is done in received_change_keep_all_no_key?
```

Comment:
Can this TODO comment be removed or addressed?

Reply:
This is a backport, and this TODO has not been addressed on the mainline.